src/store.rs

```diff
@@ -44,52 +43,81 @@ impl Store {
-            let source = PathBuf::from(output.as_str());
+            self.associate_job_with_hash(job, &item.hash().to_string())
+                .context("could not associate job with hash")
+        } else {
+            log::debug!("moving {} into store", item);
 
-            ///////////////////////////////////////////////////
-            // Step 1: Validate that the path we got is safe //
-            ///////////////////////////////////////////////////
-            for component in source.components() {
-                match component {
-                    Component::Prefix(_) | Component::RootDir =>  anyhow::bail!(
-                        "Got `{}` as an output, but absolute paths are not allowed as outputs. Remove the absolute prefix to fix this!",
-                        source.display(),
-                    ),
+            let hash = item
+                .move_into(&self.root)
+                .context("could not move item into the store")?;
 
-                    Component::ParentDir => anyhow::bail!(
-                        "Got `{}` as an output, but relative paths containing `..` are not allowed as inputs. Remove the `..` to fix this!",
-                        source.display(),
-                    ),
+            self.associate_job_with_hash(job, &hash.to_string())
+                .context("could not associate job with hash")
+        }
+    }
 
-                    Component::CurDir | Component::Normal(_) => (),
-                };
-            }
+    fn associate_job_with_hash(&mut self, job: &Job, hash: &str) -> Result<()> {
+        self.inputs_to_content.insert(job.id, hash.to_owned());
 
-            ////////////////////////////////////////////////////////////////
-            // Step 2: add the file content to the content-addressed hash //
-            ////////////////////////////////////////////////////////////////
-            hasher.update(output.as_bytes());
+        let file = std::fs::File::create(self.root.join("inputs_to_content.json"))
+            .context("failed to open job-to-content-hash mapping")?;
+        // TODO: BufWriter?
+        serde_json::to_writer(file, &self.inputs_to_content)
+            .context("failed to write job-to-content-hash mapping")
+    }
+}
+
+/// ContentAddressedItem is responsible for hashing the outputs of a job inside
+/// a workspace and (maybe) moving those outputs into the store.
+#[derive(Debug)]
+struct ContentAddressedItem<'job> {
+    workspace: Workspace,
+    job: &'job Job,
+    hash: blake3::Hash,
+}
+
+impl<'job> ContentAddressedItem<'job> {
+    /// Load all the outputs from a job and workspace combo, creating a hash
+    /// as we go.
+    fn load(job: &'job Job, workspace: Workspace) -> Result<Self> {
+        let mut hasher = blake3::Hasher::new();
 
-            let mut file = File::open(&workspace.join(output.as_str())).with_context(|| {
+        for path in job.outputs.iter().sorted() {
+            hasher.update(path.to_string_lossy().as_bytes());
```

Comment:
I'm not sure `to_string_lossy` is the right method here. There is also `to_str` and `to_os_str`. I guess it depends if you plan to support all os strings possible, or just utf8. Also, I guess, if these came from Roc, they should already be valid utf8.

Reply:
Yeah, good point. It'd probably be better to bail instead of introducing potentially-corrupt paths into the store key. I'll make it bail here instead.